File: lib/py/telepact/internal/CreateInternalFunctionRoutes.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from ..FunctionRouter import FunctionRoute
from ..Message import Message
from ..TelepactSchema import TelepactSchema
from .GetApiDefinitionsWithExamples import get_api_definitions_with_examples


def create_internal_function_routes(telepact_schema: TelepactSchema) -> dict[str, FunctionRoute]:
    async def handle_ping(_function_name: str, _request_message: Message) -> Message:
        return Message({}, {"Ok_": {}})

    async def handle_api(_function_name: str, request_message: Message) -> Message:
        request_payload = request_message.get_body_payload()
        include_internal = request_payload.get("includeInternal!") is True
        include_examples = request_payload.get("includeExamples!") is True
        api_definitions = get_api_definitions_with_examples(
            telepact_schema,
            include_internal,
        ) if include_examples else (
            telepact_schema.full if include_internal else telepact_schema.original
        )
        return Message({}, {"Ok_": {"api": api_definitions}})

    return {
        "fn.ping_": handle_ping,
        "fn.api_": handle_api,
    }

File: lib/ts/src/internal/CreateInternalFunctionRoutes.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { FunctionRoutes } from '../FunctionRouter.js';
import { Message } from '../Message.js';
import { TelepactSchema } from '../TelepactSchema.js';
import { getApiDefinitionsWithExamples } from './GetApiDefinitionsWithExamples.js';

export function createInternalFunctionRoutes(telepactSchema: TelepactSchema): FunctionRoutes {
    return {
        'fn.ping_': async (): Promise<Message> => new Message({}, { Ok_: {} }),
        'fn.api_': async (_functionName: string, requestMessage: Message): Promise<Message> => {
            const requestPayload = requestMessage.getBodyPayload();
            const includeInternal = requestPayload['includeInternal!'] === true;
            const includeExamples = requestPayload['includeExamples!'] === true;
            const apiDefinitions = includeExamples
                ? getApiDefinitionsWithExamples(telepactSchema, includeInternal)
                : includeInternal
                    ? telepactSchema.full
                    : telepactSchema.original;
            return new Message({}, { Ok_: { api: apiDefinitions } });
        },
    };
}
